### PR TITLE
Rebuild comm package

### PR DESCRIPTION
**Pull request by ftheirs**
**HDPath changed to string**
The user can select which elements are hardened. The previous implementation used an array and the hardening was hidden from the user.

- Previous: `path := []uint32{44, 118, 0, 0, 21}`
- New: `path := "m/44'/118'/0'/0/21"`

**Removed all legacy v1 and v2 functions**
- GetVersion
- GetAddressAndPubKey
- sign

**Sign function will require** 
- HRP (compatibility with EVM chains: Injective, Evmos)
- Transaction type (SignModeAmino / SignModeTextual)

**Encapsulation from responses in structures**
- VersionResponse
- AddressResponse
- SignatureResponse

**Fixed issues**
https://github.com/cosmos/ledger-cosmos-go/issues/52
https://github.com/cosmos/ledger-cosmos-go/issues/53
https://github.com/cosmos/ledger-cosmos-go/issues/54

### CODE REVIEW

File: common.go [hunk 1]
```diff
@@ -18,89 +18,83 @@ package ledger_cosmos_go
...
-func (c VersionInfo) String() string {
-	return fmt.Sprintf("%d.%d.%d", c.Major, c.Minor, c.Patch)
-}
-
-// VersionRequiredError the command is not supported by this app
-type VersionRequiredError struct {
-	Found    VersionInfo
-	Required VersionInfo
+type AddressResponse struct {
+	pubkey  []byte // 0: release | 0xFF: debug
```
Comment (raynaudoe) [inline common.go:37]: is this comment valid also for `pubkey`?

File: common.go [hunk 2]
```diff
@@ -18,89 +18,83 @@ package ledger_cosmos_go
...
+	address string
 }
 
-func (e VersionRequiredError) Error() string {
-	return fmt.Sprintf("App Version required %s - Version found: %s", e.Required, e.Found)
+type SignatureResponse struct {
+	signatureDER []byte
```
Comment (raynaudoe) [inline common.go:42]: Maybe here you need a getter for `signatureDER` if you don't want to make that field public

File: common.go [hunk 3]
```diff
@@ -18,89 +18,83 @@ package ledger_cosmos_go
...
-func CheckVersion(ver VersionInfo, req VersionInfo) error {
-	if ver.Major != req.Major {
-		if ver.Major > req.Major {
-			return nil
-		}
-		return NewVersionRequiredError(req, ver)
+// Validate HRP: Max length = 83
+// All characters must be in range [33, 126], displayable chars in Ledger devices
+func serializeHRP(hrp string) (hrpBytes []byte, err error) {
+	if len(hrp) > 83 {
```
Comment (raynaudoe) [inline common.go:52]: Avoid using magic numbers if possible, for example here you could create a `constants.go` file and add something like:
```go
const (
HRP_MAX_LENGTH = 83
// all other constant values
)
```

File: common.go [hunk 4]
```diff
@@ -18,89 +18,83 @@ package ledger_cosmos_go
...
+		return nil, errors.New("HRP len should be <= 83")
```
Comment (raynaudoe) [inline common.go:53]: also here, refer to that const

File: common.go [hunk 5]
```diff
@@ -18,89 +18,83 @@ package ledger_cosmos_go
...
 	}
 
-	if ver.Minor != req.Minor {
-		if ver.Minor > req.Minor {
-			return nil
+	hrpBytes = []byte(hrp)
+	for _, b := range hrpBytes {
+		if b < 33 || b > 126 {
```
Comment (raynaudoe) [inline common.go:58]: samething here

File: common.go [hunk 6]
```diff
@@ -18,89 +18,83 @@ package ledger_cosmos_go
...
-	message[0] = byte(len(bip32Path))
-	for index, element := range bip32Path {
-		pos := 1 + index*4
-		value := element
-		if index < hardenCount {
-			value = 0x80000000 | element
-		}
-		binary.LittleEndian.PutUint32(message[pos:], value)
+func serializePath(path string) (pathBytes []byte, err error) {
+	const HARDENED = 0x80000000
```
Comment (raynaudoe) [inline common.go:67]: also this one can be moved to constants.go

File: common.go [hunk 7]
```diff
@@ -18,89 +18,83 @@ package ledger_cosmos_go
...
-}
 
-func GetBip32bytesv2(bip44Path []uint32, hardenCount int) ([]byte, error) {
-	message := make([]byte, 20)
-	if len(bip44Path) != 5 {
-		return nil, fmt.Errorf("path should contain 5 elements")
+	pathArray := strings.Split(path, "/")
+	pathArray = pathArray[1:] // remove "m"
+
+	if len(pathArray) != 5 {
```
Comment (raynaudoe) [inline common.go:76]: same with this value

**Review by raynaudoe — APPROVED** [review]
Only a few nits!

File: common.go (hunk 2 above)
Comment (ftheirs) [reply common.go:42]: There isn't any problem if this field is public or not. At the moment, the app only delivers the signature of the transaction in DER format. It's just a clarification if the user needs to verify the signature.

**Pull request closed without merging**